Add support for CocoaPods dependencies

pepibumur opened the pull request:
Resolves https://github.com/tuist/tuist/issues/420

### Short description 📝

Adds support for setting CocoaPods dependencies to targets.

### Solution 📦
- Targets have now support for a `TargetDependency` of type `.cocoapods(path: ".")`, where path is the relative path to the directory that contains the Podfile.
- The CocoaPods nodes are part of the graph when its loaded.
- After the projects are generated, Tuist runs `pod install --repo-update` for each of the CocoaPods nodes.

> Note: Tuist does not validate that the configuration in the Podfiles is right. If the user doesn't set the project/workspace right, the generation of will fail because CocoaPods won't be able to find the project to integrate the projects into.

### Implementation 👩‍💻👨‍💻

- [x] Add models.
- [x] Include the CocoaPods nodes in the graph.
- [x] Add CocoaPods interactor utility.
- [x] Add documentation.

### Testing plan ✅
- [x] Unit tests.
- [x]  Acceptance test that generates a project that has a CocoaPods dependency, and verifies the resulting project and workspace has been configured by CocoaPods.

ollieatkinson commented:
Great work @pepibumur! 

I created a fixture and tested it out, it would be nice if we could start to infer the Podfile too (automate it). This should be done in a follow-up P/R anyway as it will need some discussion.

It would also be nice to automatically include the Podfile inside of the workspace.

---

This P/R works great!

```ruby
use_frameworks!

platform :ios, '10.0'

workspace 'Workspace.xcworkspace'

target 'App' do

    project 'App/MainApp.xcodeproj'

    pod 'FirebaseAuth'
    pod 'FirebaseCore'
    pod 'FirebaseFirestore'

end
```

```
~/tuist/development/tuist/fixtures/ios_app_with_cocoapods [cocoapods ?] $ swift run tuist generate --path .
Generating workspace Workspace.xcworkspace
Generating project Framework2
Generating project MainApp
Generating project Framework1
/Users/oat01/.gem/ruby/2.5.5/gems/cocoapods-1.7.3
Installing CocoaPods dependencies defined in /Users/oat01/tuist/development/tuist/fixtures/ios_app_with_cocoapods/Podfile
Analyzing dependencies
Downloading dependencies
Installing BoringSSL-GRPC (0.0.3)
Installing FirebaseAuth (6.2.1)
Installing FirebaseAuthInterop (1.0.0)
Installing FirebaseCore (6.1.0)
Installing FirebaseFirestore (1.4.2)
Installing GTMSessionFetcher (1.2.2)
Installing GoogleUtilities (6.2.3)
Installing Protobuf (3.9.0)
Installing gRPC-C++ (0.0.9)
Installing gRPC-Core (1.21.0)
Installing leveldb-library (1.20)
Installing nanopb (0.3.901)
Generating Pods project
Integrating client project
Sending stats
Pod installation complete! There are 3 dependencies from the Podfile and 12 total pods installed.
✅ Success: Project generated.
Total time taken: 56.130s
```

ollieatkinson commented on Sources/TuistGenerator/Utils/CocoaPodsInteractor.swift:
```diff
@@ -0,0 +1,102 @@
+                throw CocoaPodsInteractorError.cocoapodsNotFound
+            }
+
+            command.append(contentsOf: ["install", "--project-directory=\(node.path.pathString)", "--repo-update"])
```
it might be a good idea to pass through `--repo-update` since you might not always want to run this when running `tuist generate`

e.g. 

```
$ swift run tuist generate --path fixtures/ios_app_with_cocoapods -Xpod --repo-update

pepibumur replied:
Even better 😬. I've changed the implementation a bit:
- By default, it runs the command without the `--repo-update` argument.
- If CocoaPods fails because it can't find the version, it re-runs the command with the `--repo-update` argument.

I think the simpler we keep Tuist's interface, the better. If it fails because the repo of specs is outdated, I think Tuist should handle that automatically.

kwridan commented on Sources/TuistGenerator/Utils/CocoaPodsInteractor.swift:
```diff
@@ -0,0 +1,102 @@
+    /// - Parameter graph: Project graph.
+    /// - Throws: An error if the installation of the pods fails.
+    func install(graph: Graphing) throws {
+        let canUseBundler = canUseCocoaPodsThroughBundler()
```
worth adding a guard here to save a bit of time from running those shell calls if there aren't any cocoapod dependencies :)

```swift
guard !graph.cocoapods.isEmpty else { 
  return 
}
```

kwridan commented on Sources/TuistGenerator/Utils/CocoaPodsInteractor.swift:
```diff
@@ -0,0 +1,102 @@
+    /// - Returns: True if Bundler can execute CocoaPods.
+    func canUseCocoaPodsThroughBundler() -> Bool {
+        do {
+            try system.runAndPrint(["bundle", "show", "cocoapods"])
```
is there a way to suppress the output?

Otherwise for a project without cocoa pods we will get this in the console :/

```
Could not locate Gemfile or .bundle/ directory
```

pepibumur replied:
Actually it should just be run. Good catch @kwridan

ollieatkinson commented:
I had a thought on this last night, since we are not creating any explicit dependency on a manifest level for the project, does it make sense for a target dependency type cocoapods rather than something global?

pepibumur commented:
> I had a thought on this last night, since we are not creating any explicit dependency on a manifest level for the project, does it make sense for a target dependency type cocoapods rather than something global?

You are totally right, but I think having a more global dependency, something like `ProjectDependency`, would confuse the users: *What's the difference between a target dependency and a project dependency? Why do we have both?*

I assume most of the developers will have the dependencies defined for the final product target *(i.e. the app)*, in which case it's fine defining it as a dependency of that target.

--

@kwridan / @ollieatkinson I addressed your comments and:
- Updated the CHANGELOG.
- Added an acceptance test.
- Revisit the install logic to run update the specs repo if the specs can't be found.

Could you have another look? 👀

ollieatkinson commented:
That makes sense @pepibumur - it also opens up the possibility for creating explicit dependencies on individual cocoapods down the line. Since we know about the project name, target and dependency information it should be fairly trivial using `Graph` to be able to generate a Podfile.

ollieatkinson commented on Sources/TuistCore/Utils/Printer.swift:
```diff
@@ -38,14 +38,14 @@ public class Printer: Printing {
 
     public func print(success: String) {
         let writer = InteractiveWriter.stdout
-        writer.write("✅ Success: ", inColor: .green, bold: true)
```
What's the reason for removing these emoji's?

pepibumur replied:
I removed it because the color and the prefix is enough to denote the type of message being printed.

kwridan replied:
There's a subtle difference when using `tuist` directly or via `tuistenv` - the colours are sadly lost in the latter case

ollieatkinson commented on docs/usage/3-dependencies.mdx:
````diff
@@ -76,6 +78,31 @@ It defines a dependency with a pre-compiled library. It allows specifying the pa
+
+```
+
+```
````
Empty code block

ollieatkinson commented on features/generate.feature:
```diff
@@ -160,3 +160,10 @@ Scenario: The project is an iOS application with multiple configurations (ios_ap
     Then the scheme Framework2 has a build setting CUSTOM_FLAG with value "Target.Beta" for the configuration Beta
     Then the scheme Framework2 has a build setting CUSTOM_FLAG with value "Release" for the configuration Release
+
+Scenario: The project is an iOS application with CocoaPods dependencies (ios_app_with_pods)
```
:+1:
